### PR TITLE
feat(worker-service): move the jobs processing logic to the new service

### PR DESCRIPTION
**Pull request by LetItRock**
## What change does this PR introduce?

In this PR I've moved the jobs processing logic to the `worker-service `. At this point the code is duplicated in two places, but I'll move the shared code to the reusable usecases library in the next PR (I don't want this to be even bigger). The tests in the `worker-service ` require some changes and they don't work.

The changes:
- `WorkflowQueueService` was renamed to `QueueService`, moved to `application-generic`, and cleared from the worker's code
- the API app now uses the `WorkflowQueueProducerService` that extends `QueueService`
- the `worker-service ` now uses the `WorkflowQueueService`that extends `QueueService` which is the jobs consumer that has the processing logic
- moved a lot of use-cases to `worker-service `
- marked API code that needs to be revisited

### Why was this change needed?

We want to move the jobs processing logic into a separate service that will be easier to dynamically scale based on the queue size.

### Other information (Screenshots)

https://user-images.githubusercontent.com/2607232/227517965-74402db5-db52-4042-80b9-8cf44398b456.mov

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -30,6 +30,20 @@
...
+        "run-script",
+        "start"
+      ],
+      "runtimeExecutable": "npm",
+      "cwd": "${workspaceFolder}/apps/jobs-processing-service",
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "type": "pwa-node"
+    },
```
Comment (LetItRock) [inline .vscode/launch.json:46]: VS Code config for debugging the `jobs-processing-service`

File: apps/api/src/app/events/e2e/delay-events.e2e.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ describe('Trigger event - Delay triggered events - /v1/events/trigger (POST)', f
   let subscriber: SubscriberEntity;
   let subscriberService: SubscribersService;
   const jobRepository = new JobRepository();
-  let workflowQueueService: WorkflowQueueService;
+  let workflowQueueProducerService: WorkflowQueueProducerService;
```
Comment (LetItRock) [inline apps/api/src/app/events/e2e/delay-events.e2e.ts:28]: renamed this queue service as now it's just the jobs producer queue

File: apps/api/src/app/events/services/chat-service/chat.factory.ts [hunk 3]
```diff
@@ -23,3 +23,5 @@ export class ChatFactory implements IChatFactory {
     }
   }
 }
+
+// TODO remove?
```
Comment (LetItRock) [inline apps/api/src/app/events/services/chat-service/chat.factory.ts:27]: marked code with `// TODO remove?` which I'll revisit in the next PR and will check if this code can be removed

File: apps/jobs-processing-service/package.json [hunk 4]
```diff
@@ -23,15 +23,55 @@
     "test:e2e": "cross-env TS_NODE_COMPILER_OPTIONS='{\"strictNullChecks\": false}' TZ=UTC NODE_ENV=test E2E_RUNNER=true mocha --timeout 10000 --require ts-node/register --exit --file e2e/setup.ts src/**/*.e2e.ts"
   },
   "dependencies": {
+    "@aws-sdk/client-s3": "^3.234.0",
```
Comment (LetItRock) [inline apps/jobs-processing-service/package.json:26]: added required deps

File: apps/jobs-processing-service/src/.env.development [hunk 5]
```diff
@@ -1,11 +1,31 @@
 NODE_ENV=dev
 PORT=3004
 
+# Novu
+MAX_NOVU_INTEGRATION_MAIL_REQUESTS=300
+# NOVU_EMAIL_INTEGRATION_API_KEY=
+
+# Storage Service
+# STORAGE_SERVICE=
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/.env.development:9]: novu integration usecase will be moved to shared usecases lib, caching and storage could be moved to `application-generic`

File: apps/jobs-processing-service/src/app/shared/commands/authenticated.command.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { IsNotEmpty } from 'class-validator';
+
+import { BaseCommand } from './base.command';
+
+export abstract class AuthenticatedCommand extends BaseCommand {
+  @IsNotEmpty()
+  public readonly userId: string;
+}
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/shared/commands/authenticated.command.ts:8]: the base commands classes could be moved to the reusable usecases lib

File: apps/jobs-processing-service/src/app/shared/distributed-lock/distributed-lock.service.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    tls: process.env.REDIS_TLS as ConnectionOptions,
+  };
+};
+
+interface ILockOptions {
+  resource: string;
+  ttl: number;
+}
+
+export class DistributedLockService {
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/shared/distributed-lock/distributed-lock.service.ts:34]: this service could be also moved to `application-generic`

File: apps/jobs-processing-service/src/app/shared/logs/create-log.usecase.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { Injectable } from '@nestjs/common';
+import { LogEntity, LogRepository } from '@novu/dal';
+
+import { CreateLogCommand } from './create-log.command';
+
+@Injectable()
+export class CreateLog {
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/shared/logs/create-log.usecase.ts:7]: this service could be also moved to shared usecases lib

File: apps/jobs-processing-service/src/app/shared/performance/performance.service.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+export interface IMark {
+  id: string;
+}
+
+interface IMeasure {
+  name: string;
+  duration: number;
+}
+
+export class PerformanceService {
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/shared/performance/performance.service.ts:27]: this service could be also moved to `application-generic`

File: apps/jobs-processing-service/src/app/shared/shared.module.ts [hunk 10]
```diff
@@ -53,6 +67,17 @@ const dalService = new DalService();
...
+function getStorageServiceClass() {
+  switch (process.env.STORAGE_SERVICE) {
+    case 'GCS':
+      return GCSStorageService;
+    case 'AZURE':
+      return AzureBlobStorageService;
+    default:
+      return S3StorageService;
+  }
+}
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/shared/shared.module.ts:79]: the storage service could be also moved to `application-generic`

File: apps/jobs-processing-service/src/app/workflow/services/workflow-queue.service.ts [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+interface IJobData {
+  _id: string;
+  _environmentId: string;
+  _organizationId: string;
+  _userId: string;
+}
+
+@Injectable()
+export class WorkflowQueueService extends QueueService<IJobData> {
```
Comment (LetItRock) [inline apps/jobs-processing-service/src/app/workflow/services/workflow-queue.service.ts:27]: the workflow jobs queue service with processing logic

**Review by ainouzgali — APPROVED** [review]
So much work🚀
Except for the copy/moving existing logic, I did not notice anything else at the moment to comment on.  
Looking forward to the end result. Looks good

File: apps/worker-service/src/.env.production
[the remark this answers is not in the document]
Comment (LetItRock) [reply apps/worker-service/src/.env.production:1]: already added it in the next PR ;)

**Pull request merged**